scripts/ASTImportTest.sh

```diff
@@ -1,105 +1,190 @@
+        print_used_commands "$(pwd)" "${export_command[*]} > expected.json" "${import_command[*]}"
+        return 1
+    fi
+
+    # compare expected and obtained ast's
+    if ! diff_files expected.json obtained.json
+    then
+        printError "❌ ERROR: AST reimport failed for ${sol_file}"
+        if [[ $EXIT_ON_ERROR == 1 ]]
         then
-            # For investigating, use exit 1 here so the script stops at the
-            # first failing test
-            # exit 1
-            FAILED=$((FAILED + 1))
-            echo -e "ERROR: AST reimport failed for input file $nth_input_file"
-            echo
-            echo "Compiler stderr:"
-            cat ./stderr.txt
-            echo
-            echo "Compiler stdout:"
-            cat ./obtained.json
+            print_used_commands "$(pwd)" "${export_command[*]}" "${import_command[*]}"
             return 1
         fi
-        DIFF="$(diff expected.json obtained.json)"
-        if [ "$DIFF" != "" ]
-        then
-            if [ "$DIFFVIEW" == "" ]
-            then
-                echo -e "ERROR: JSONS differ for $1: \n $DIFF \n"
-                echo "Expected:"
-                cat ./expected.json
-                echo "Obtained:"
-                cat ./obtained.json
-            else
-                # Use user supplied diff view binary
-                $DIFFVIEW expected.json obtained.json
-            fi
-            FAILED=$((FAILED + 1))
-            return 2
-        fi
-        TESTED=$((TESTED + 1))
-        rm expected.json obtained.json
-        rm -f stderr.txt
+        FAILED=$((FAILED + 1))
+    fi
+    TESTED=$((TESTED + 1))
+}
+
+# function tests whether exporting and importing again is equivalent.
+# Results are recorded by adding to FAILED or UNCOMPILABLE.
+# Also, in case of a mismatch a diff is printed
+# Expected parameters:
+# $1 name of the file to be exported and imported
+# $2 any files needed to do so that might be in parent directories
+function test_import_export_equivalence {
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
+    local OUTPUT
+    local SOLC_RC
+
+    set +e
+    OUTPUT=$("$SOLC" --bin "${input_files[@]}" 2>&1)
+    SOLC_RC=$?
+    set -e
+
+    # if input files where compilable with success
+    if (( SOLC_RC == 0 ))
+    then
+        case "$IMPORT_TEST_TYPE" in
+            ast) ast_import_export_equivalence "${sol_file}" "${input_files[@]}" ;;
+            *) fail "Unknown import test type '${IMPORT_TEST_TYPE}'. Aborting." ;;
+        esac
     else
-        # echo "contract $solfile could not be compiled "
         UNCOMPILABLE=$((UNCOMPILABLE + 1))
+        # solc will return exit code 2, if it was terminated by an uncaught exception.
+        if (( SOLC_RC == 2 ))
+        then
+          # Ignore all UnimplementedFeatureError exceptions.
+          if [[ "$OUTPUT" != *"UnimplementedFeatureError"* ]]
```

Comment:
It's obvious that you're ignoring these errors, the thing you should say is why you're doing it.

Is that because isoltest catches them and prints them as normal errors while solc does not?

Reply:
I just realised now why we had so many uncompilable tests if we use the syntax tests here. The current version of this script is just using `solc` on the input files https://github.com/ethereum/solidity/blob/73e7b844dce88717151dd0f11f137a024ac1c810/scripts/ASTImportTest.sh#L43

During the refactoring I saw that some errors are only detectable, if the code generator is really generating code, e.g. if `--bin` was selected as output mode. So I changed the compilability test, to let also the code generator run. But of course, there is no need to let the code generator run for the AST import/export tests. So I changed this now to`--ast-compact-json` as the compilability check. This compilability check will be different for the evm asm import/export tests, because here we need to run the code generator. 

However, because of this I will remove that check for `UnimplementedFeatureError` from this PR now. It is even possible that this extra check is not needed anymore in the context of the evm asm import/export, because those are not using the syntax tests anymore.